### PR TITLE
Fix normalization/stretch classes for upstream mpl change

### PR DESCRIPTION
**Pull request by larrybradley**
matplotlib 3.2.0 introduced a change in behavior such that NaN values get mapped the colormap "bad" value instead of the "under" value (https://github.com/matplotlib/matplotlib/pull/14257).  The problem is that  several of our Stretch classes currently generate NaN values (**that are not present in the input data**) when `clip=False`, e.g.,:

```python
>>> from astropy.visualization import SqrtStretch
>>> data = [-2., -1, 0, 1, 2]
>>> SqrtStretch()(data, clip=False)
array([       nan,        nan, 0.        , 1.        , 1.41421356])
```

For `mpl >= 3.2`, the effect is that good image values are masked as "bad" values when using these stretch classes with our `ImageNormalize` class:

```python
import matplotlib.pyplot as plt
from astropy.visualization import ImageNormalize, SqrtStretch
np.random.seed(123)
data = np.random.uniform(-10, 10, size=(25, 25))
norm = ImageNormalize(data, stretch=SqrtStretch(), vmin=-5, clip=False)
plt.imshow(data, norm=norm)
plt.colorbar()
```

`mpl 3.1.3`:
mpl-313

`mpl 3.2.1`:
mpl-321

The white pixels above are where `data < 0`, thus the `SqrtStretch` produces NaN, which now get masked as bad values by mpl >= 3.2.  This is not the expected/desired result.

After offline discussion with @astrofrog, I've added a `invalid` keyword to the  `SqrtStretch`, `LogStretch`, and `PowerStretch` classes.  The `invalid` keyword is used to assign newly-generated NaN values; it should be set to map to the matplotlib colormap "under" value (i.e., any finite value < 0).  For the stretch classes, the default is `None`, which means the NaN values are not replaced.

I've also added the `invalid` keyword to the `ImageNormalize` class and the `simple_norm` function, both with a default value of -1.  With this default value, the normalization produces the expected result with mpl (for all versions).

This PR is a followup to #10177.

### CODE REVIEW

File: docs/visualization/normalization.rst [hunk 1]
```diff
@@ -213,9 +213,19 @@ the data and the interval and stretch objects:
...
 As shown above, the colorbar ticks are automatically adjusted.
 
-The input image to :class:`~astropy.visualization.mpl_normalize.ImageNormalize`
-is typically the one to be displayed, so there is a convenience function
-:func:`~astropy.visualization.mpl_normalize.imshow_norm` to ease this use case:
+Please note that one should not use ``ax.imshow(norm(image))`` because
+the colorbar ticks marks will represent normalized image values (on a
+linear scale), not the actual image values.  Also, the image displayed
+by ``ax.imshow(norm(image))`` is not exactly equivalent to
+``ax.imshow(image, norm=norm)`` if the image contains `NaN` or `inf`
```
Comment (bsipocz) [inline docs/visualization/normalization.rst:220]:
```suggestion
``ax.imshow(image, norm=norm)`` if the image contains ``NaN`` or ``inf``
```

**Comment by pllim** [comment]
So this doesn't need a backport to LTS?

HTML failure is real:
```
/home/circleci/project/docs/visualization/normalization.rst:216:py:obj reference target not found: NaN
```

**Comment by larrybradley** [comment]
@pllim It would be great to have this backported, but I've added keywords to several classes.  Is that allowed for backports?

File: docs/visualization/normalization.rst (hunk 1 above)
Comment (larrybradley) [reply docs/visualization/normalization.rst]: Thanks!

**Comment by bsipocz** [comment]
>  It would be great to have this backported, but I've added keywords to several classes. Is that allowed for backports?

If the default behaviour is the same, I think yes, this is fine as part of a compatibility bugfix. At least definitely better than not being able to reproduce mpl behaviour on LTS with the upcoming mpl version

**Comment by bsipocz** [comment]
(changelog can mention the mpl version, and/or even placed into a `Other Changes and Additions` section if needed).

Let's see what @astrofrog thinks

File: astropy/visualization/mpl_normalize.py [hunk 2]
```diff
@@ -58,10 +58,17 @@ class ImageNormalize(Normalize):
     clip : bool, optional
         If `True`, data values outside the [0:1] range are clipped to
         the [0:1] range.
+    invalid : `None` or float, option
```
Comment (astrofrog) [inline astropy/visualization/mpl_normalize.py:61]: optional

File: astropy/visualization/mpl_normalize.py [hunk 3]
```diff
@@ -217,6 +257,14 @@ def simple_norm(data, stretch='linear', power=1.0, asinh_a=0.1, min_cut=None,
     log_a : float, optional
         The log index for ``stretch='log'``. The default is 1000.
 
+    invalid : `None` or float, option
```
Comment (astrofrog) [inline astropy/visualization/mpl_normalize.py:260]: optional

File: astropy/visualization/mpl_normalize.py (hunk 2 above)
Comment (larrybradley) [reply astropy/visualization/mpl_normalize.py]: Done.

File: astropy/visualization/mpl_normalize.py (hunk 3 above)
Comment (larrybradley) [reply astropy/visualization/mpl_normalize.py]: Done.  (plus others)

**Comment by bsipocz** [comment]
Thanks @larrybradley!

**Pull request merged**